### PR TITLE
Fix #4327: SmallModulesFor module split style

### LINKED ISSUE #4327: Intermediate SplitStyle mode
I have been trying to use `snowpack` with scala.js new split modules in development mode.

This is really meant for development mode where I want to change code in my application and want to see the results as soon as possible in the browser. A tool like `snowpack` is able to detect what changed and only load those

I tried the setting:
```scala
    scalaJSLinkerConfig ~= (_.withModuleSplitStyle(ModuleSplitStyle.SmallestModules)),
```
And that will create modules for each class and indeed only the few that change will be updated and captured by `snowpack`

However for my test application there is a large set of modules, in my case 3172 files including e.g. one for each scala class in use, or  dependencies like `scalajs-react`

```
ls -al target/scala-2.13/seqexec_fastopt/
...
-rw-------     1 cquiroz  staff    11819 Dec 11 14:28 japgolly.scalajs.react.extra.router.RoutingRules.js
...
-rw-------     1 cquiroz  staff    25248 Dec 11 17:25 web.client.table.package$.js
...
-rw-------     1 cquiroz  staff     1190 Dec 11 14:28 java.lang.ThreadLocal.js
...
```

Though `snowpack` only detects the changed files, the sheer amount of files means it needs to check a lot of files and it also puts pressure on the browser as well (needing to track thousands of modules).

What would be great (I have no idea how involved this is) is to have modules grouped, e.g. all from dependency X into an `X.js` and keep the application classes split, this would let the dev tool to load application files as they change, while caching dependencies grouped in "bigger" files

I don't if this is feasible but I'll open this to discuss

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/ModuleSplitter.scala [hunk 1]
```diff
@@ -177,11 +177,16 @@ final class ModuleSplitter private (analyzer: ModuleAnalyzer) {
...
-  def maxSplitter(): ModuleSplitter =
-    new ModuleSplitter(new MaxModuleAnalyzer())
+  def smallestModules(): ModuleSplitter =
+    new ModuleSplitter(new SmallestModulesAnalyzer())
+
+  def fewestModules(): ModuleSplitter =
+    new ModuleSplitter(new FewestModulesAnalyzer())
+
+  def smallModulesFor(packages: List[String]): ModuleSplitter =
+    // TODO: What if a prefix is not valid?
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/ModuleSplitter.scala:188]: @sjrd I guess we don't have another choice but to duplicate the validation logic in `ClassName` for a `java.lang.String` in `linkter.interface`?

**Comment by gzm0** [comment]
TODO:
- [ ] validate package names
- [ ] add tests to Jenkinsfile

File: linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleSplitStyle.scala [hunk 2]
```diff
@@ -27,13 +27,26 @@ object ModuleSplitStyle {
...
+   *
+   *  * Make modules as small as possible for all classes in any of [[packages]]
+   *    (or subpackages thereof).
+   *  * Make as few modules as possible for everything else (while not including
+   *    unnecessary code).
+   *
+   *  @note In order to avoid ambiguity between packages and classes (and a
+   *     simple interface), selecting individual classes is not supported.
+   */
+  case class SmallModulesFor(packages: List[String]) extends ModuleSplitStyle
```
Comment (cquiroz) [inline linker-interface/shared/src/main/scala/org/scalajs/linker/interface/ModuleSplitStyle.scala:40]: This is really cool.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/ModuleSplitter.scala (hunk 1 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/ModuleSplitter.scala:188]: I guess the `// TODO` line should be removed now.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/SmallestModulesAnalyzer.scala [hunk 3]
```diff
@@ -33,146 +32,21 @@ private[modulesplitter] final class SmallestModulesAnalyzer extends ModuleAnalyz
...
-
-      node
+    protected def emitModule(moduleIndex: Int, classNames: List[ClassName]): Unit = {
+      val repr = ModuleIDs.representativeClass(classNames)
+      val id = ModuleIDs.forClassName(info.publicModuleDependencies.keySet, repr)
+      moduleIndexToID(moduleIndex) = id
     }
   }
+
 }
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/SmallestModulesAnalyzer.scala:52]: Spurious new blank line here.

File: linker/shared/src/test/scala/org/scalajs/linker/SmallModulesForSplittingTest.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        module.classDefs.map(_.name.name)
+      }
+
+      assertEquals(List[ClassName]("foo.A"), moduleClasses("foo.A"))
+      assertEquals(List[ClassName]("foo.C"), moduleClasses("foo.C"))
+      assertEquals(List(MainTestClassName), moduleClasses("main"))
+
+      /* Expect two additional modules on for each:
```
Comment (sjrd) [inline linker/shared/src/test/scala/org/scalajs/linker/SmallModulesForSplittingTest.scala:88]: Was it supposed to read as
```suggestion
      /* Expect two additional modules, one for each:
```
?

**Review by sjrd — CHANGES_REQUESTED** [review]
Reviewed everything except `SmallModulesForAnalyzer.scala` (yes, I know it's the most important file of this PR ^^)

**Review by sjrd — COMMENTED** [review]
And here is for `SmallModulesForAnalyzer`.

It's very nice that there is so little code in there, and that it can actually reuse code from the other two analyzers! :)

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/SmallModulesForAnalyzer.scala [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    protected def emitModule(moduleIndex: Int, classNames: List[ClassName]): Unit = {
+      // Does this strongly connected component contain a target class.
+      val targetNames = classNames.filter(clazz => packages.exists(inPackage(clazz, _)))
+
+      if (targetNames.nonEmpty) {
+        val repr = ModuleIDs.representativeClass(targetNames)
+        val id = ModuleIDs.forClassName(info.publicModuleDependencies.keySet, repr)
+        reprToModuleID(repr) = id
+        classNames.foreach(targetClassToRepr(_) = repr)
```
[the remark this answers is not in the document]
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/SmallModulesForAnalyzer.scala:83]: I switched this to a for loop. Somewhat in line with the codingstyle that side-effecting statements must be on their own line.

**Review by sjrd — APPROVED** [review]
Needs a rebase, and apparently there are a few scalastyle issues. Otherwise LGTM.

**Comment by sjrd** [comment]
There's a compiler error in `OptimizerTest.scala`. Probably caused by rebasing.

**Review by sjrd — APPROVED** [review]

**Pull request merged**